Added CharField field type. Added to Model::insertBatch for data type…

rowan08 opened the pull request:
… CHAR.

Issue: https://github.com/davidbrownza/CORMAP/issues/10

rowan08 committed:
Added CharField field type. Added to Model::insertBatch for data type CHAR.
Issue: https://github.com/davidbrownza/CORMAP/issues/10

rowan08 committed:
Redid CharField, including maxLength validation.
Issue: https://github.com/davidbrownza/cormap/issues/10

davidbrownza commented on Field.h:
```diff
@@ -125,4 +126,38 @@ class TextField: public Field {
+            }
+            return newVal;            
+        }
 };
```
This is new to me too, but there should be a newline at the end of the file: https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

davidbrownza replied:
PS I know it was already missing the newline, but just add it in while you are at it. There are a few other files like this as well.

davidbrownza commented on Field.h:
```diff
@@ -125,4 +126,38 @@ class TextField: public Field {
+            return maxLength;
+        }
+        
+        string checkMaxLen(string newVal, int MaxLenVal) {
```
`MaxLenVal` should start with a lowercase: `maxLenVal`.

davidbrownza replied:
Actually, for consistency, this should be called `checkMaxLength` and the first parameter should be `newValue`. The last parameter is not required as it is already a class attribute.

davidbrownza commented on Field.h:
```diff
@@ -125,4 +126,38 @@ class TextField: public Field {
+        string getValue() {
+            return fieldValue;
+        }
+        int getMaxLen() {
```
Consistent spacing between methods. I have been using 1 newline between each method (I think - maybe just check the rest of the file to make sure), rather stick to that.

davidbrownza replied:
This method should be `getMaxLength`

davidbrownza commented on Field.h:
```diff
@@ -125,4 +126,38 @@ class TextField: public Field {
+        int maxLength;
+        
+    public:
+        CharField(string name, int maxLen, string defaultValue="", bool primary=false, bool unique=false, bool nullable=true):
```
Second parameter should be `maxLength`. I prefer using the full word e.g. `defaultValue` instead of `defaultVal`. Keeping these sorts of things consistent lets people guess correctly.

rowan08 replied:
Here `maxLength` has already been declared as a private variable above (line 135). I think you had the same problem with the `Field` class where the final parameter is set as `bool autoF=false` for the member variable `autoFilled`. We will need to decide what to do with these cases in future since they will probably be quite common:
https://stackoverflow.com/questions/10250016/should-i-use-the-same-name-for-a-member-variable-and-a-function-parameter-in-c

davidbrownza replied:
Hmmm I see that. Seems there is quite a lot of my stuff that needs to be refactored as well. The `_isNull` also breaks consistency.

Okay, this is what I think we should do:
1. all private member variables get an underscore in front of them (someone using this class can't access those variables directly anyway, so I would rather they were ugly than the method parameters).
1. We avoid use of `this->` and `this.`
1. We make the method signatures as pretty as possible, meaning full length words and consistency throughout.

We may need to look through my existing code and make sure it conforms or update it if it doesn't. Don't worry about that for this task though, rather just write your class with these rules.

davidbrownza commented on Model.cpp:
```diff
@@ -37,6 +37,14 @@ TextField* Model::textField(string name, string defaultValue, bool primary, bool
     return f;
 }
 
+CharField* Model::charField(string name, int maxLen, string defaultValue, bool primary, bool unique, bool nullable)
```
Second parameter should be `maxLength`.

davidbrownza commented on Model.h:
```diff
@@ -25,6 +25,7 @@ class Model
         IntegerField* integerField(string name, int defaultValue=0, bool primary=false, bool unique=false, bool nullable=true, bool autoInc=false);
         FloatField* floatField(string name, double defaultValue=0, bool primary=false, bool unique=false, bool nullable=true);
         TextField* textField(string name, string defaultValue="", bool primary=false, bool unique=false, bool nullable=true);
+        CharField* charField(string name, int maxLen, string defaultValue="", bool primary=false, bool unique=false, bool nullable=true);
```
Parameter name again :-P

davidbrownza commented on test.cpp:
```diff
@@ -8,6 +8,7 @@ class Protein: public Model
 	    Protein(): Model("Proteins") {}
 	    
 	    IntegerField * proteinID = integerField ("Protein_ID");
+        CharField * proteinChain = charField ("Protein_Chain", 5); // Added for charField test
```
Looks like there is an issue with the spacing here.

rowan08 replied:
There were still tab characters in this file. Apparently, I didn't think `test.cpp`  needed to be included when I switched tabs for spaces. It has been fixed now.

davidbrownza reviewed (changes requested):
Not bad, mostly pedantic stuff.

rowan08 committed:
Added new line at the end of each file, where missing.
Changed parameter names to be 'prettier' and added underscores to private variables (for the charField class only).

Code for this issue has been tested by adding a charField instance to test.cpp with two test cases. 1) where the fieldValue entered is shorter than maxLength - runs fine, and 2) where fieldValue is longer than maxLength - raises exception.
Issue: https://github.com/davidbrownza/CORMAP/issues/10

davidbrownza commented on Field.h:
```diff
@@ -125,4 +126,39 @@ class TextField: public Field {
+        
+    public:
+        CharField(string name, int maxLength, string defaultValue="", bool primary=false, bool unique=false, bool nullable=true):
+            Field(name, CHAR, primary, unique, nullable, maxLength) 
```
Why are you passing maxLength to the parent class? The last parameter can be left blank in this case.

davidbrownza reviewed (changes requested):
One comment. These files also still need new lines on the end.

rowan08 committed:
Removed charField maxLength argument from being passed to Field parent class.
Ran compile.sh again to test, which ran fine.

Issue: https://github.com/davidbrownza/CORMAP/issues/10

rowan08 committed:
Merge branch 'master' into add-char-field
